packages/ifpack2/test/unit_tests/Ifpack2_UnitTestOverlappingRowMatrix.cpp

```diff
@@ -109,6 +110,145 @@ typedef Tpetra::global_size_t GST;
+                const auto A_val = A_row.value(iEntry);
+                lsum += A_val * X_lcl(A_row.colidx(iEntry),0);
+              }, A_x);
+            Y_lcl(lclRow,0) = A_x;
+            
+          });//end parallel_for TeamThreadRange
+      });//end parallel_for "residual-vector"
+  } else {
+    // MultiVector case
+    // Kernel interior shamelessly horked from Ifpack2_Details_ScaledDampedResidual_def.hpp
+    Kokkos::parallel_for("reduced-mv-multivector",policy,KOKKOS_LAMBDA(const team_member& dev) {
+        // NOTE: It looks like I should be able to get this data up above, but if I try to
+        // we get internal compiler errors.  Who knew that gcc tried to "gimplify"?
+        const LO numVectors = static_cast<LO>(X_lcl.extent(1));
+        Kokkos::parallel_for(Kokkos::TeamThreadRange (dev, 0, rows_per_team),[&] (const LO& loop) {
+            const LO lclRow = static_cast<LO> (dev.league_rank ()) * rows_per_team + loop;
+            
+            if (lclRow >= numLocalRows) {
+              return;
+            }
+            const auto A_row = A_lcl.rowConst(lclRow);
+            const LO row_length = static_cast<LO> (A_row.length);
+            for(LO v=0; v<numVectors; v++) {
+              residual_value_type A_x = KAT::zero ();          
+              
+              Kokkos::parallel_reduce(Kokkos::ThreadVectorRange (dev, row_length), [&] (const LO iEntry, residual_value_type& lsum) {
+                  const auto A_val = A_row.value(iEntry);
+                  lsum += A_val * X_lcl(A_row.colidx(iEntry),v);
+                }, A_x);
+              Y_lcl(lclRow,v) = A_x;
+              
+            }//end for numVectors
+          });//end parallel_for TeamThreadRange
+      });//end parallel_for "residual-multivector"
+  }// end else                                     
+}// end reducedMatvec
+
+
+    
+/***********************************************************************************/
+template<class OverlappedMatrixClass, class MultiVectorClass>
+void reducedMatvec(const OverlappedMatrixClass & A,
+                   const MultiVectorClass & X,
+                   const int userExtNumRows,
+                   MultiVectorClass & Y) {
+  using crs_matrix_type = Tpetra::CrsMatrix<typename OverlappedMatrixClass::scalar_type,
+    typename OverlappedMatrixClass::local_ordinal_type,
+    typename OverlappedMatrixClass::global_ordinal_type,
+    typename OverlappedMatrixClass::node_type>;
+
+  // Assumes that X& Y are sufficiently overlapped for this to work
+  RCP<const crs_matrix_type> undA = Teuchos::rcp_dynamic_cast<const crs_matrix_type>(A.getUnderlyingMatrix());
+  RCP<const crs_matrix_type> extA = Teuchos::rcp_dynamic_cast<const crs_matrix_type>(A.getExtMatrix());
+
+  auto undA_lcl = undA->getLocalMatrix ();
+  auto extA_lcl = extA->getLocalMatrix ();
+  auto X_lcl = X.getLocalViewDevice ();
```

Comment:
We should make sure that X and Y are correctly sync'd -- otherwise, we'll be assuming UVM implicitly.

Reply:
For this test that's not true --- everything should be happening on the device.  In general, we should definitely check sync states first.